tests/external_executor/executors/logging_app.py

```diff
@@ -0,0 +1,100 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the Apache 2.0 License.
+import grpc
+import json
+import urllib
+
+from loguru import logger as LOG
+
+# pylint: disable=import-error
+import kv_pb2 as KV
+
+# pylint: disable=import-error
+import http_pb2 as HTTP
+
+# pylint: disable=import-error
+import kv_pb2_grpc as Service
+
+# pylint: disable=no-name-in-module
+from google.protobuf.empty_pb2 import Empty as Empty
+
+
+class LoggingExecutor:
+    def __init__(self, ccf_node, credentials):
+        self.ccf_node = ccf_node
+        self.credentials = credentials
+
+    def do_post(self, kv_stub, table, request, response):
+        body = json.loads(request.body)
+        kv_stub.Put(
+            KV.KVKeyValue(
+                table=table,
+                key=body["id"].to_bytes(8, "big"),
+                value=body["msg"].encode("utf-8"),
+            )
+        )
+        response.status_code = HTTP.HttpStatusCode.OK
+        header = response.headers.add()
+        header.field = "content-type"
+        header.value = "application/json"
+        response.body = json.dumps(True).encode("utf-8")
+
+    def do_get(self, kv_stub, table, request, response):
+        query_args = urllib.parse.parse_qs(request.query)
+        msg_id = int(query_args["id"][0])
+        result = kv_stub.Get(
+            KV.KVKey(
+                table=table,
+                key=msg_id.to_bytes(8, "big"),
+            )
+        )
+
+        if not result.HasField("optional"):
+            response.status_code = HTTP.HttpStatusCode.BAD_REQUEST
+            response.body = f"No such record: {msg_id}"
+            return
+
+        response.status_code = HTTP.HttpStatusCode.OK
+        response.body = json.dumps(
+            {"msg": result.optional.value.decode("utf-8")}
+        ).encode("utf-8")
+
+    def run_loop(self, terminate_event):
+        target_uri = f"{self.ccf_node.get_public_rpc_host()}:{self.ccf_node.get_public_rpc_port()}"
+        with grpc.secure_channel(
+            target=target_uri,
+            credentials=self.credentials,
+        ) as channel:
+            stub = Service.KVStub(channel)
+
+            while not (terminate_event.is_set()):
+                request_description_opt = stub.StartTx(Empty())
```

Comment:
This is just a question rather than request to change.
It looks like it pulls requests very actively. Is it effective enough for us? Or are we going to make some changes like making the executer a http server?

Reply:
The eventual plan (may be possible after #4537), is to make `StartTx` a streaming RPC. Since we don't currently support streaming, we get a similar result by polling in a loop. The result should be similar to making the executor a server (gets _sent_ requests to execute, as soon as they arrive, rather than having to manually ask for them), without requiring that the executors are reachable as a server (ie - have some unique network name).

We could sleep or backoff in this polling loop, trading off responsiveness/latency for reducing load. If we need to keep the polling long-term, I think we'll do that, but since its a temporary inefficiency I think we're happy to keep this sample simple (if inefficient).

Reply:
Thanks. I understood it. Making `StartTx` sounds nice also because it only requires gRPC which we already have.